### PR TITLE
Renesas: Smartbond: Update counter driver to work with PM

### PR DESCRIPTION
**Pull request by ioannis-karachalios**

### CODE REVIEW

File: drivers/counter/counter_smartbond_timer.c [hunk 1]
```diff
@@ -47,15 +64,120 @@ struct counter_smartbond_config {
...
+	uint8_t trigger = counter_smartbond_pdc_trigger_get(dev);
+
+	if (atomic_test_and_set_bit(data->pm_policy_state_flag,
+				COUNTER_SMARTBOND_PM_POLICY_STATE_PDC) == 0) {
+
+		data->pdc_idx = da1469x_pdc_add(trigger, MCU_PDC_MASTER_M33, PDC_XTAL_EN);
+		__ASSERT_NO_MSG(data->pdc_idx >= 0);
+
+		da1469x_pdc_set(data->pdc_idx);
+		da1469x_pdc_set(data->pdc_idx);
```
Comment (ydamigos) [inline drivers/counter/counter_smartbond_timer.c:144]:
```suggestion
		da1469x_pdc_ack(data->pdc_idx);
```

Comment (ioannis-karachalios) [reply drivers/counter/counter_smartbond_timer.c]: Fixed

**Review by ydamigos — APPROVED** [review]

**Review by blauret — DISMISSED** [review]

**Comment by blauret** [comment]
@andrzej-kaczmarek ping

File: drivers/counter/counter_smartbond_timer.c [hunk 2]
```diff
@@ -47,15 +64,120 @@ struct counter_smartbond_config {
 	LOG_INSTANCE_PTR_DECLARE(log);
 };
 
+#if defined(CONFIG_PM_DEVICE) || defined(CONFIG_PM_DEVICE_RUNTIME)
+static void counter_smartbond_pm_policy_state_lock_get(struct counter_smartbond_data *data)
+{
+	if (atomic_test_and_set_bit(data->pm_policy_state_flag,
+				COUNTER_SMARTBOND_PM_POLICY_STATE_DRV) == 0) {
```
Comment (andrzej-kaczmarek) [inline drivers/counter/counter_smartbond_timer.c:71]: atomic_test_and_set_bit returns bool so you should not use explicit comparison with int

File: drivers/counter/counter_smartbond_timer.c [hunk 3]
```diff
@@ -47,15 +64,120 @@ struct counter_smartbond_config {
...
+		pdc_trigger = MCU_PDC_TRIGGER_TIMER2;
+		break;
+	case (uint32_t)TIMER3:
+		pdc_trigger = MCU_PDC_TRIGGER_TIMER3;
+		break;
+	case (uint32_t)TIMER4:
+		pdc_trigger = MCU_PDC_TRIGGER_TIMER4;
+		break;
+	default:
+		return -EINVAL;
```
Comment (andrzej-kaczmarek) [inline drivers/counter/counter_smartbond_timer.c:121]: return value type is uint8_t

File: drivers/counter/counter_smartbond_timer.c [hunk 4]
```diff
@@ -22,11 +25,25 @@ LOG_MODULE_REGISTER(counter_timer, CONFIG_COUNTER_LOG_LEVEL);
...
+	COUNTER_SMARTBOND_PM_POLICY_STATE_PDC,
+};
+
 struct counter_smartbond_data {
 	counter_alarm_callback_t callback;
 	void *user_data;
 	uint32_t guard_period;
 	uint32_t freq;
+#if defined(CONFIG_PM_DEVICE) || defined(CONFIG_PM_DEVICE_RUNTIME)
+	ATOMIC_DEFINE(pm_policy_state_flag, 2);
```
Comment (andrzej-kaczmarek) [inline drivers/counter/counter_smartbond_timer.c:44]: do we need a bitmask here? both options seem to be mutually exclusive so a single flag that indicates whether timer is pd_tim or pd_sys would be enough.

**Review by andrzej-kaczmarek — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_smartbond_timer.c (hunk 2 above)
Comment (ioannis-karachalios) [reply drivers/counter/counter_smartbond_timer.c]: Fixed where needed

File: drivers/counter/counter_smartbond_timer.c (hunk 3 above)
Comment (ioannis-karachalios) [reply drivers/counter/counter_smartbond_timer.c]: Fixed

File: drivers/counter/counter_smartbond_timer.c (hunk 4 above)
Comment (ioannis-karachalios) [reply drivers/counter/counter_smartbond_timer.c]: You are right. Fixed.

**Review by blauret — APPROVED** [review]

**Review by andrzej-kaczmarek — APPROVED** [review]

**Pull request merged**